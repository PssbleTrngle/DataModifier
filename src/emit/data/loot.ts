import { Acceptor } from '@pssbletrngle/pack-resolver'
import { encodeId, Id, IdInput, NormalizedId, prefix } from '../../common/id.js'
import {
   CommonTest,
   IngredientInput,
   IngredientTest,
   Predicate,
   resolveIngredientTest,
} from '../../common/ingredient.js'
import { resolveIDTest } from '../../common/predicates.js'
import RegistryLookup from '../../loader/registry/index.js'
import { TagRegistryHolder } from '../../loader/tags.js'
import { Logger } from '../../logger.js'
import { createLootEntry, LootItemInput, replaceItemInTable } from '../../parser/lootTable.js'
import LootTableRule from '../rule/lootTable.js'
import { EmptyLootEntry, LootModifier, LootTable, LootTableSchema } from '../../schema/data/loot.js'
import CustomEmitter from '../custom.js'
import { ClearableEmitter, RegistryProvider } from '../index.js'
import RuledEmitter from '../ruled.js'

export const EMPTY_LOOT_TABLE: LootTable = {
   type: 'minecraft:empty',
   pools: [],
}

export const EMPTY_LOOT_MODIFIER: LootModifier = {
   type: 'noop',
}

type LootTableTest = Readonly<{
   id?: CommonTest<NormalizedId>
   output?: IngredientTest
}>

export interface LootRules {
   replaceOutput(from: IngredientTest, to: LootItemInput, additionalTests?: LootTableTest): void

   removeOutput(from: IngredientTest, additionalTests?: LootTableTest): void

   add(id: IdInput, value: LootTable): void

   disable(test: LootTableTest): void

   block(id: IdInput): void

   addModifier<T extends LootModifier>(id: IdInput, value: T): void

   disabledModifier(id: IdInput): void
}

export default class LootTableEmitter implements LootRules, ClearableEmitter {
   private readonly customTables = new CustomEmitter<LootTable>(this.tablePath)
   private readonly customModifiers = new CustomEmitter<LootModifier>(this.modifierPath)

   private readonly ruled = new RuledEmitter<LootTable, LootTableRule>(
      this.logger,
      this.lootTables,
      this.tablePath,
      EMPTY_LOOT_TABLE,
      id => this.customTables.has(id)
   )

   constructor(
      private readonly logger: Logger,
      private readonly lootTables: RegistryProvider<LootTable>,
      private readonly tags: TagRegistryHolder,
      private readonly lookup: () => RegistryLookup
   ) {}

   private tablePath(id: Id) {
      return `data/${id.namespace}/loot_tables/${id.path}.json`
   }

   private modifierPath(id: Id) {
      return `data/${id.namespace}/loot_modifiers/${id.path}.json`
   }

   clear() {
      this.customTables.clear()
      this.customModifiers.clear()
      this.ruled.clear()
   }

   async emit(acceptor: Acceptor) {
      await Promise.all([
         this.ruled.emit(acceptor),
         this.customTables.emit(acceptor),
         this.customModifiers.emit(acceptor),
      ])
   }

   resolveIngredientTest(test: IngredientTest) {
      return resolveIngredientTest(test, this.tags, this.lookup())
   }

   private resolveLootTableTest(test: LootTableTest) {
      const id: Predicate<Id>[] = []
      const output: Predicate<IngredientInput>[] = []

      if (test.id) id.push(resolveIDTest(test.id))
      if (test.output) output.push(this.resolveIngredientTest(test.output))

      return { id, output }
   }

   add(id: IdInput, value: LootTable): void {
      this.customTables.add(id, LootTableSchema.parse(value))
   }

   disable(test: LootTableTest): void {
      const predicates = this.resolveLootTableTest(test)
      this.ruled.addRule(new LootTableRule(predicates.id, predicates.output, () => null))
   }

   replaceOutput(from: IngredientTest, to: LootItemInput, additionalTests: LootTableTest = {}): void {
      const predicates = this.resolveLootTableTest(additionalTests)
      const outputPredicate = this.resolveIngredientTest(from)
      const replacer = replaceItemInTable(outputPredicate, createLootEntry(to, this.lookup()))
      this.ruled.addRule(new LootTableRule(predicates.id, [outputPredicate, ...predicates.output], replacer))
   }

   removeOutput(from: IngredientTest, additionalTests?: LootTableTest) {
      this.replaceOutput(from, EmptyLootEntry, additionalTests)
   }

   block(id: IdInput) {
      this.add(prefix(id, 'blocks/'), {
         type: 'minecraft:block',
         pools: [
            {
               rolls: 1,
               entries: [
                  {
                     type: 'minecraft:item',
                     name: encodeId(id),
                  },
               ],
               conditions: [
                  {
                     condition: 'minecraft:survives_explosion',
                  },
               ],
            },
         ],
      })
   }

   disabledModifier(id: IdInput) {
      this.addModifier(id, EMPTY_LOOT_MODIFIER)
   }

   addModifier<T extends LootModifier>(id: IdInput, value: T) {
      this.customModifiers.add(id, value)
   }
}
