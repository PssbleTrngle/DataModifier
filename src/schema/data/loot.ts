import zod from 'zod'

const NumberProviderSchema = zod.union([
   zod.number(),
   zod.object({
      type: zod.string().optional(),
   }),
])

const LootConditionSchema = zod.object({
   condition: zod.string(),
})

const LootFunctionSchema = zod.object({
   function: zod.string(),
})

const LootEntryBaseSchema = zod
   .object({
      type: zod.string(),
      conditions: zod.array(LootConditionSchema).optional(),
      functions: zod.array(LootFunctionSchema).optional(),
   })
   .passthrough()

export type LootEntryBase = zod.infer<typeof LootEntryBaseSchema>

const LootEntryAlternativeSchema = zod.object({
   type: zod.literal('minecraft:alternatives'),
   children: zod.array(LootEntryBaseSchema),
})

const LootEntryItemSchema = zod.object({
   type: zod.literal('minecraft:item'),
   name: zod.string(),
})

const LootEntryTagSchema = zod.object({
   type: zod.literal('minecraft:tag'),
   name: zod.string(),
   expand: zod.boolean().optional(),
})

const LootEntryEmptySchema = zod.object({
   type: zod.literal('minecraft:empty'),
})

export const EmptyLootEntry: LootEntry = {
   type: 'minecraft:empty',
}

const LootEntryReferenceSchema = zod.object({
   type: zod.literal('minecraft:loot_table'),
   name: zod.string(),
})

export const LootEntrySchema = LootEntryBaseSchema.and(
   zod.discriminatedUnion('type', [
      LootEntryAlternativeSchema,
      LootEntryItemSchema,
      LootEntryTagSchema,
      LootEntryReferenceSchema,
      LootEntryEmptySchema,
   ])
)

export type LootEntry = zod.infer<typeof LootEntrySchema>

export const LootPoolSchema = zod.object({
   rolls: NumberProviderSchema,
   bonus_rolls: NumberProviderSchema.optional(),
   entries: zod.array(LootEntryBaseSchema),
   conditions: zod.array(LootConditionSchema).optional(),
   functions: zod.array(LootFunctionSchema).optional(),
})

export type LootPool = zod.infer<typeof LootPoolSchema>

export const LootTableSchema = zod.object({
   type: zod.string().optional(),
   pools: zod.array(LootPoolSchema).default([]),
})

export type LootTable = zod.infer<typeof LootTableSchema>

export function extendLootEntry(base: LootEntryBase): LootEntry {
   return LootEntrySchema.parse(base)
}
