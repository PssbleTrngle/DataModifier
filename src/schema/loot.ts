import zod from 'zod'

const NumberProviderSchema = zod.union([
   zod.number(),
   zod.object({
      type: zod.string(),
   }),
])

const LootConditionSchema = zod.object({
   condition: zod.string(),
})

const LootFunctionSchema = zod.object({
   function: zod.string(),
})

const LootEntryBaseSchema = zod
   .object({
      type: zod.string(),
      conditions: zod.array(LootConditionSchema).optional(),
      functions: zod.array(LootFunctionSchema).optional(),
   })
   .passthrough()

export type LootEntryBase = zod.infer<typeof LootEntryBaseSchema>

export const LootEntryAlternativeSchema = zod.object({
   type: zod.literal('minecraft:alternatives'),
   children: zod.array(LootEntryBaseSchema),
})

export type LootEntryAlternative = zod.infer<typeof LootEntryAlternativeSchema>

export const LootEntryItemSchema = zod.object({
   type: zod.literal('minecraft:item'),
   name: zod.string(),
})

export type LootEntryItem = zod.infer<typeof LootEntryItemSchema>

export const LootEntryTagSchema = zod.object({
   type: zod.literal('minecraft:tag'),
   name: zod.string(),
   expand: zod.boolean().optional(),
})

export type LootEntryTag = zod.infer<typeof LootEntryTagSchema>

export const LootEntryEmptySchema = zod.object({
   type: zod.literal('minecraft:empty'),
})

export type LootEntryEmpty = zod.infer<typeof LootEntryEmptySchema>

export const LootEntryReferenceSchema = zod.object({
   type: zod.literal('minecraft:loot_table'),
   name: zod.string(),
})

export type LootEntryReference = zod.infer<typeof LootEntryReferenceSchema>

export const LootEntrySchema = LootEntryBaseSchema.and(
   zod.discriminatedUnion('type', [
      LootEntryAlternativeSchema,
      LootEntryItemSchema,
      LootEntryTagSchema,
      LootEntryReferenceSchema,
      LootEntryEmptySchema,
   ])
)

export type LootEntry = zod.infer<typeof LootEntrySchema>

export const LootPoolSchema = zod.object({
   rolls: NumberProviderSchema,
   bonus_rolls: NumberProviderSchema.optional(),
   entries: zod.array(LootEntryBaseSchema),
})

export type LootPool = zod.infer<typeof LootPoolSchema>

export const LootTableSchema = zod.object({
   type: zod.string(),
   pools: zod.array(LootPoolSchema).default([]),
})

export type LootTable = zod.infer<typeof LootTableSchema>

export function extendLootEntry(base: LootEntryBase): LootEntry {
   return LootEntrySchema.parse(base)
}
