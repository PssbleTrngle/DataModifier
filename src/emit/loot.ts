import { Logger } from '../logger.js'
import TagsLoader from '../loader/tags.js'
import { EmptyLootEntry, LootTable, LootTableSchema } from '../schema/loot.js'
import { Acceptor } from '@pssbletrngle/pack-resolver'
import RuledEmitter from './ruled.js'
import CustomEmitter from './custom.js'
import { Id, IdInput, NormalizedId } from '../common/id.js'
import LootTableRule from '../rule/lootTable.js'
import { RegistryProvider } from './index.js'
import { CommonTest, IngredientInput, IngredientTest, Predicate, resolveIngredientTest } from '../common/ingredient.js'
import { resolveIDTest } from '../common/predicates.js'
import { createLootEntry, LootItemInput, replaceItemInTable } from '../parser/lootTable.js'

export const EMPTY_LOOT_TABLE: LootTable = {
   type: 'minecraft:empty',
   pools: [],
}

type LootTableTest = Readonly<{
   id?: CommonTest<NormalizedId>
   output?: IngredientTest
}>

export interface LootRules {
   replaceOutput(from: IngredientTest, to: LootItemInput, additionalTests?: LootTableTest): void

   removeOutput(from: IngredientTest, additionalTests?: LootTableTest): void

   addLootTable(id: IdInput, value: LootTable): void

   disableLootTable(test: LootTableTest): void
}

export default class LootTableEmitter implements LootRules {
   private readonly custom = new CustomEmitter<LootTable>(this.lootPath)

   private readonly ruled = new RuledEmitter<LootTable, LootTableRule>(
      this.logger,
      this.lootTables,
      this.lootPath,
      EMPTY_LOOT_TABLE,
      id => this.custom.has(id)
   )

   constructor(
      private readonly logger: Logger,
      private readonly lootTables: RegistryProvider<LootTable>,
      private readonly tags: TagsLoader
   ) {}

   private lootPath(id: Id) {
      return `data/${id.namespace}/loot_tables/${id.path}.json`
   }

   async emit(acceptor: Acceptor) {
      await Promise.all([this.ruled.emit(acceptor), this.custom.emit(acceptor)])
   }

   resolveIngredientTest(test: IngredientTest) {
      return resolveIngredientTest(test, this.tags)
   }

   private resolveLootTableTest(test: LootTableTest) {
      const id: Predicate<Id>[] = []
      const output: Predicate<IngredientInput>[] = []

      if (test.id) id.push(resolveIDTest(test.id))
      if (test.output) output.push(this.resolveIngredientTest(test.output))

      return { id, output }
   }

   addLootTable(id: IdInput, value: LootTable): void {
      this.custom.add(id, LootTableSchema.parse(value))
   }

   disableLootTable(test: LootTableTest): void {
      const predicates = this.resolveLootTableTest(test)
      this.ruled.addRule(new LootTableRule(predicates.id, predicates.output, () => null))
   }

   replaceOutput(from: IngredientTest, to: LootItemInput, additionalTests: LootTableTest = {}): void {
      const predicates = this.resolveLootTableTest(additionalTests)
      const outputPredicate = this.resolveIngredientTest(from)
      const replacer = replaceItemInTable(outputPredicate, createLootEntry(to))
      this.ruled.addRule(new LootTableRule(predicates.id, [outputPredicate, ...predicates.output], replacer))
   }

   removeOutput(from: IngredientTest, additionalTests?: LootTableTest) {
      this.replaceOutput(from, EmptyLootEntry, additionalTests)
   }

   clear() {
      this.custom.clear()
      this.ruled.clear()
   }
}
