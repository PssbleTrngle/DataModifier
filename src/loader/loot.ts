import { encodeId, Id } from '../common/id.js'
import { tryCatching } from '../error.js'
import { LootTable, LootTableSchema } from '../schema/loot.js'
import { JsonLoader } from './index.js'
import { Logger } from '../logger.js'

export default class LootTableLoader extends JsonLoader<LootTable> {
   protected parse(logger: Logger, json: unknown, id: Id): LootTable | null {
      return tryCatching(logger.group(`error parsing loot table ${encodeId(id)}`), () => {
         return LootTableSchema.parse(json)
      })
   }
}
