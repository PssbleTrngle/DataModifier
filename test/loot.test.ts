import createTestAcceptor from './mock/TestAcceptor.js'
import { LootEntrySchema, LootTableSchema } from '../src/schema/data/loot.js'
import { EMPTY_LOOT_TABLE } from '../src/emit/data/loot.js'
import setupLoader from './shared/loaderSetup.js'

const { logger, loader } = setupLoader({ include: ['data/*/loot_tables/**/*.json', 'data/*/tags/**/*.json'] })

afterEach(() => {
   loader.clear()
})

describe('loading of loot tables', () => {
   it('loads loot tables without errors', async () => {
      expect(logger.warn).not.toHaveBeenCalled()
      expect(logger.error).not.toHaveBeenCalled()
   })
})

describe('loot tables output replacements', () => {
   it('removes outputs', async () => {
      const acceptor = createTestAcceptor()

      loader.loot.removeOutput('#minecraft:iron_ores')

      await loader.emit(acceptor)

      expect(acceptor.jsonAt('data/minecraft/loot_tables/blocks/iron_ore.json')).toMatchSnapshot(
         'modified deepslate_iron_ore loot table'
      )
      expect(acceptor.jsonAt('data/minecraft/loot_tables/blocks/deepslate_iron_ore.json')).toMatchSnapshot(
         'modified iron_ore loot table'
      )
   })

   it('replaces outputs with additional tests', async () => {
      const acceptor = createTestAcceptor()

      loader.loot.replaceOutput('#forge:ingots/iron', { tag: 'forge:ingots/lead' }, { id: /minecraft:entities\/.+/ })

      await loader.emit(acceptor)

      expect(acceptor.paths().length).toBe(4)

      expect(acceptor.jsonAt('data/minecraft/loot_tables/entities/husk.json')).toMatchSnapshot(
         'modified husk loot table'
      )
      expect(acceptor.jsonAt('data/minecraft/loot_tables/entities/iron_golem.json')).toMatchSnapshot(
         'modified iron_golem loot table'
      )
      expect(acceptor.jsonAt('data/minecraft/loot_tables/entities/zombie.json')).toMatchSnapshot(
         'modified zombie loot table'
      )
      expect(acceptor.jsonAt('data/minecraft/loot_tables/entities/zombie_villager.json')).toMatchSnapshot(
         'modified zombie_villager loot table'
      )
   })

   it('keeps extended loot entry properties', async () => {
      const acceptor = createTestAcceptor()

      loader.loot.replaceOutput('farmersdelight:rice', { item: 'minecraft:apple' })

      await loader.emit(acceptor)

      expect(acceptor.jsonAt('data/farmersdelight/loot_tables/blocks/wild_rice.json')).toMatchSnapshot(
         'modified wild rice loot table'
      )
   })
})

describe('loot table removal', () => {
   it('removes loot table with id filter', async () => {
      const acceptor = createTestAcceptor()

      loader.loot.disable({
         id: /minecraft:.*oak_log/,
      })

      await loader.emit(acceptor)

      expect(acceptor.paths().length).toBe(4)

      expect(acceptor.jsonAt('data/minecraft/loot_tables/blocks/oak_log.json')).toMatchObject(EMPTY_LOOT_TABLE)
      expect(acceptor.jsonAt('data/minecraft/loot_tables/blocks/stripped_oak_log.json')).toMatchObject(EMPTY_LOOT_TABLE)
      expect(acceptor.jsonAt('data/minecraft/loot_tables/blocks/dark_oak_log.json')).toMatchObject(EMPTY_LOOT_TABLE)
      expect(acceptor.jsonAt('data/minecraft/loot_tables/blocks/stripped_dark_oak_log.json')).toMatchObject(
         EMPTY_LOOT_TABLE
      )
   })

   it('removes loot table with output filter', async () => {
      const acceptor = createTestAcceptor()

      loader.loot.disable({
         output: '#minecraft:logs',
      })

      await loader.emit(acceptor)

      expect(acceptor.paths()).toMatchSnapshot('loot tables containing any log')
   })
})

it('creates custom loot tables', async () => {
   const acceptor = createTestAcceptor()

   const lootTable = LootTableSchema.parse({
      type: 'minecraft:block',
      pools: [
         {
            rolls: 4,
            entries: [
               LootEntrySchema.parse({
                  type: 'minecraft:alternatives',
                  children: [
                     LootEntrySchema.parse({
                        type: 'minecraft:item',
                        name: 'minecraft:diamond',
                     }),
                  ],
               }),
               LootEntrySchema.parse({
                  type: 'minecraft:tag',
                  name: 'minecraft:logs',
               }),
            ],
         },
      ],
   })

   loader.loot.add('example:custom', lootTable)

   await loader.emit(acceptor)

   expect(acceptor.jsonAt('data/example/loot_tables/custom.json')).toMatchSnapshot('parsed loot table')
   expect(acceptor.jsonAt('data/example/loot_tables/custom.json')).toMatchObject(lootTable)
})
