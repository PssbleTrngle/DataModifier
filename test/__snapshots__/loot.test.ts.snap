// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`creates custom loot tables: parsed loot table 1`] = `
{
  "pools": [
    {
      "entries": [
        {
          "children": [
            {
              "name": "minecraft:diamond",
              "type": "minecraft:item",
            },
          ],
          "type": "minecraft:alternatives",
        },
        {
          "name": "minecraft:logs",
          "type": "minecraft:tag",
        },
      ],
      "rolls": 4,
    },
  ],
  "type": "minecraft:block",
}
`;

exports[`loot table removal removes loot table with output filter: loot tables containing any log 1`] = `
[
  "data/minecraft/loot_tables/blocks/acacia_log.json",
  "data/minecraft/loot_tables/blocks/acacia_wood.json",
  "data/minecraft/loot_tables/blocks/birch_log.json",
  "data/minecraft/loot_tables/blocks/birch_wood.json",
  "data/minecraft/loot_tables/blocks/crimson_hyphae.json",
  "data/minecraft/loot_tables/blocks/crimson_stem.json",
  "data/minecraft/loot_tables/blocks/dark_oak_log.json",
  "data/minecraft/loot_tables/blocks/dark_oak_wood.json",
  "data/minecraft/loot_tables/blocks/jungle_log.json",
  "data/minecraft/loot_tables/blocks/jungle_wood.json",
  "data/minecraft/loot_tables/blocks/oak_log.json",
  "data/minecraft/loot_tables/blocks/oak_wood.json",
  "data/minecraft/loot_tables/blocks/spruce_log.json",
  "data/minecraft/loot_tables/blocks/spruce_wood.json",
  "data/minecraft/loot_tables/blocks/stripped_acacia_log.json",
  "data/minecraft/loot_tables/blocks/stripped_acacia_wood.json",
  "data/minecraft/loot_tables/blocks/stripped_birch_log.json",
  "data/minecraft/loot_tables/blocks/stripped_birch_wood.json",
  "data/minecraft/loot_tables/blocks/stripped_crimson_hyphae.json",
  "data/minecraft/loot_tables/blocks/stripped_crimson_stem.json",
  "data/minecraft/loot_tables/blocks/stripped_dark_oak_log.json",
  "data/minecraft/loot_tables/blocks/stripped_dark_oak_wood.json",
  "data/minecraft/loot_tables/blocks/stripped_jungle_log.json",
  "data/minecraft/loot_tables/blocks/stripped_jungle_wood.json",
  "data/minecraft/loot_tables/blocks/stripped_oak_log.json",
  "data/minecraft/loot_tables/blocks/stripped_oak_wood.json",
  "data/minecraft/loot_tables/blocks/stripped_spruce_log.json",
  "data/minecraft/loot_tables/blocks/stripped_spruce_wood.json",
  "data/minecraft/loot_tables/blocks/stripped_warped_hyphae.json",
  "data/minecraft/loot_tables/blocks/stripped_warped_stem.json",
  "data/minecraft/loot_tables/blocks/warped_hyphae.json",
  "data/minecraft/loot_tables/blocks/warped_stem.json",
  "data/minecraft/loot_tables/chests/pillager_outpost.json",
  "data/minecraft/loot_tables/chests/spawn_bonus_chest.json",
  "data/minecraft/loot_tables/chests/village/village_taiga_house.json",
]
`;

exports[`loot tables output replacements keeps extended loot entry properties: modified wild rice loot table 1`] = `
{
  "pools": [
    {
      "conditions": [
        {
          "block": "farmersdelight:wild_rice",
          "condition": "minecraft:block_state_property",
          "properties": {
            "half": "lower",
          },
        },
        {
          "condition": "minecraft:location_check",
          "offsetY": 1,
          "predicate": {
            "block": {
              "block": "farmersdelight:wild_rice",
              "state": {
                "half": "upper",
              },
            },
          },
        },
      ],
      "entries": [
        {
          "children": [
            {
              "conditions": [
                {
                  "condition": "minecraft:match_tool",
                  "predicate": {
                    "tag": "forge:shears",
                  },
                },
              ],
              "name": "farmersdelight:wild_rice",
              "type": "minecraft:item",
            },
            {
              "conditions": [
                {
                  "condition": "minecraft:survives_explosion",
                },
              ],
              "name": "minecraft:apple",
              "type": "minecraft:item",
            },
          ],
          "type": "minecraft:alternatives",
        },
      ],
      "rolls": 1,
    },
    {
      "conditions": [
        {
          "block": "farmersdelight:wild_rice",
          "condition": "minecraft:block_state_property",
          "properties": {
            "half": "upper",
          },
        },
        {
          "condition": "minecraft:location_check",
          "offsetY": -1,
          "predicate": {
            "block": {
              "block": "farmersdelight:wild_rice",
              "state": {
                "half": "lower",
              },
            },
          },
        },
      ],
      "entries": [
        {
          "children": [
            {
              "conditions": [
                {
                  "condition": "minecraft:match_tool",
                  "predicate": {
                    "tag": "forge:shears",
                  },
                },
              ],
              "name": "farmersdelight:wild_rice",
              "type": "minecraft:item",
            },
            {
              "conditions": [
                {
                  "condition": "minecraft:survives_explosion",
                },
              ],
              "name": "minecraft:apple",
              "type": "minecraft:item",
            },
          ],
          "type": "minecraft:alternatives",
        },
      ],
      "rolls": 1,
    },
  ],
  "type": "minecraft:block",
}
`;

exports[`loot tables output replacements removes outputs: modified deepslate_iron_ore loot table 1`] = `
{
  "pools": [
    {
      "bonus_rolls": 0,
      "entries": [
        {
          "children": [
            {
              "conditions": [
                {
                  "condition": "minecraft:match_tool",
                  "predicate": {
                    "enchantments": [
                      {
                        "enchantment": "minecraft:silk_touch",
                        "levels": {
                          "min": 1,
                        },
                      },
                    ],
                  },
                },
              ],
              "type": "minecraft:empty",
            },
            {
              "functions": [
                {
                  "enchantment": "minecraft:fortune",
                  "formula": "minecraft:ore_drops",
                  "function": "minecraft:apply_bonus",
                },
                {
                  "function": "minecraft:explosion_decay",
                },
              ],
              "name": "minecraft:raw_iron",
              "type": "minecraft:item",
            },
          ],
          "type": "minecraft:alternatives",
        },
      ],
      "rolls": 1,
    },
  ],
  "type": "minecraft:block",
}
`;

exports[`loot tables output replacements removes outputs: modified iron_ore loot table 1`] = `
{
  "pools": [
    {
      "bonus_rolls": 0,
      "entries": [
        {
          "children": [
            {
              "conditions": [
                {
                  "condition": "minecraft:match_tool",
                  "predicate": {
                    "enchantments": [
                      {
                        "enchantment": "minecraft:silk_touch",
                        "levels": {
                          "min": 1,
                        },
                      },
                    ],
                  },
                },
              ],
              "type": "minecraft:empty",
            },
            {
              "functions": [
                {
                  "enchantment": "minecraft:fortune",
                  "formula": "minecraft:ore_drops",
                  "function": "minecraft:apply_bonus",
                },
                {
                  "function": "minecraft:explosion_decay",
                },
              ],
              "name": "minecraft:raw_iron",
              "type": "minecraft:item",
            },
          ],
          "type": "minecraft:alternatives",
        },
      ],
      "rolls": 1,
    },
  ],
  "type": "minecraft:block",
}
`;

exports[`loot tables output replacements replaces outputs with additional tests: modified husk loot table 1`] = `
{
  "pools": [
    {
      "bonus_rolls": 0,
      "entries": [
        {
          "functions": [
            {
              "add": false,
              "count": {
                "max": 2,
                "min": 0,
                "type": "minecraft:uniform",
              },
              "function": "minecraft:set_count",
            },
            {
              "count": {
                "max": 1,
                "min": 0,
                "type": "minecraft:uniform",
              },
              "function": "minecraft:looting_enchant",
            },
          ],
          "name": "minecraft:sand",
          "type": "minecraft:item",
        },
      ],
      "rolls": 1,
    },
    {
      "bonus_rolls": 0,
      "conditions": [
        {
          "condition": "minecraft:killed_by_player",
        },
        {
          "chance": 0.025,
          "condition": "minecraft:random_chance_with_looting",
          "looting_multiplier": 0.01,
        },
      ],
      "entries": [
        {
          "name": "forge:ingots/lead",
          "type": "minecraft:tag",
        },
        {
          "name": "minecraft:carrot",
          "type": "minecraft:item",
        },
        {
          "functions": [
            {
              "conditions": [
                {
                  "condition": "minecraft:entity_properties",
                  "entity": "this",
                  "predicate": {
                    "flags": {
                      "is_on_fire": true,
                    },
                  },
                },
              ],
              "function": "minecraft:furnace_smelt",
            },
          ],
          "name": "minecraft:potato",
          "type": "minecraft:item",
        },
      ],
      "rolls": 1,
    },
  ],
  "type": "minecraft:entity",
}
`;

exports[`loot tables output replacements replaces outputs with additional tests: modified iron_golem loot table 1`] = `
{
  "pools": [
    {
      "bonus_rolls": 0,
      "entries": [
        {
          "functions": [
            {
              "add": false,
              "count": {
                "max": 2,
                "min": 0,
                "type": "minecraft:uniform",
              },
              "function": "minecraft:set_count",
            },
          ],
          "name": "minecraft:poppy",
          "type": "minecraft:item",
        },
      ],
      "rolls": 1,
    },
    {
      "bonus_rolls": 0,
      "entries": [
        {
          "functions": [
            {
              "add": false,
              "count": {
                "max": 5,
                "min": 3,
                "type": "minecraft:uniform",
              },
              "function": "minecraft:set_count",
            },
          ],
          "name": "forge:ingots/lead",
          "type": "minecraft:tag",
        },
      ],
      "rolls": 1,
    },
  ],
  "type": "minecraft:entity",
}
`;

exports[`loot tables output replacements replaces outputs with additional tests: modified zombie loot table 1`] = `
{
  "pools": [
    {
      "bonus_rolls": 0,
      "entries": [
        {
          "functions": [
            {
              "add": false,
              "count": {
                "max": 2,
                "min": 0,
                "type": "minecraft:uniform",
              },
              "function": "minecraft:set_count",
            },
            {
              "count": {
                "max": 1,
                "min": 0,
                "type": "minecraft:uniform",
              },
              "function": "minecraft:looting_enchant",
            },
          ],
          "name": "minecraft:rotten_flesh",
          "type": "minecraft:item",
        },
      ],
      "rolls": 1,
    },
    {
      "bonus_rolls": 0,
      "conditions": [
        {
          "condition": "minecraft:killed_by_player",
        },
        {
          "chance": 0.025,
          "condition": "minecraft:random_chance_with_looting",
          "looting_multiplier": 0.01,
        },
      ],
      "entries": [
        {
          "name": "forge:ingots/lead",
          "type": "minecraft:tag",
        },
        {
          "name": "minecraft:carrot",
          "type": "minecraft:item",
        },
        {
          "functions": [
            {
              "conditions": [
                {
                  "condition": "minecraft:entity_properties",
                  "entity": "this",
                  "predicate": {
                    "flags": {
                      "is_on_fire": true,
                    },
                  },
                },
              ],
              "function": "minecraft:furnace_smelt",
            },
          ],
          "name": "minecraft:potato",
          "type": "minecraft:item",
        },
      ],
      "rolls": 1,
    },
  ],
  "type": "minecraft:entity",
}
`;

exports[`loot tables output replacements replaces outputs with additional tests: modified zombie_villager loot table 1`] = `
{
  "pools": [
    {
      "bonus_rolls": 0,
      "entries": [
        {
          "functions": [
            {
              "add": false,
              "count": {
                "max": 2,
                "min": 0,
                "type": "minecraft:uniform",
              },
              "function": "minecraft:set_count",
            },
            {
              "count": {
                "max": 1,
                "min": 0,
                "type": "minecraft:uniform",
              },
              "function": "minecraft:looting_enchant",
            },
          ],
          "name": "minecraft:rotten_flesh",
          "type": "minecraft:item",
        },
      ],
      "rolls": 1,
    },
    {
      "bonus_rolls": 0,
      "conditions": [
        {
          "condition": "minecraft:killed_by_player",
        },
        {
          "chance": 0.025,
          "condition": "minecraft:random_chance_with_looting",
          "looting_multiplier": 0.01,
        },
      ],
      "entries": [
        {
          "name": "forge:ingots/lead",
          "type": "minecraft:tag",
        },
        {
          "name": "minecraft:carrot",
          "type": "minecraft:item",
        },
        {
          "functions": [
            {
              "conditions": [
                {
                  "condition": "minecraft:entity_properties",
                  "entity": "this",
                  "predicate": {
                    "flags": {
                      "is_on_fire": true,
                    },
                  },
                },
              ],
              "function": "minecraft:furnace_smelt",
            },
          ],
          "name": "minecraft:potato",
          "type": "minecraft:item",
        },
      ],
      "rolls": 1,
    },
  ],
  "type": "minecraft:entity",
}
`;
